### PR TITLE
[MRG] SQLA 14 migration continued

### PR DESCRIPTION
**Pull request by ShmuelTreiger**
Working on #903 

In SQLA 2.0, using session.query is being replaced by session.execute(select()). They aren't deprecating it, but they won't be supporting it in the future. This branch will attempt to remove session.query from the codebase. It is as of yet, incomplete.

**Backend checklist**
- [ ] Formatted my code by running `autoflake -r -i --remove-all-unused-imports src && isort . && black .` in `app/backend`
- [ ] Added tests for any new code or added a regression test if fixing a bug
- [ ] All tests pass
- [ ] Run the backend locally and it works
- [ ] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

### CODE REVIEW

File: app/backend/src/couchers/couchers_select.py
Comment (kalvdans) [inline app/backend/src/couchers/couchers_select.py]: It would be fantastic if we could use subqueries for this to make the select query construction independent of the actual session.

File: app/backend/src/couchers/models.py
Comment (kalvdans) [inline app/backend/src/couchers/models.py]: Move this function out of models.py to eliminate the circular import problem (i.e. do no select in models.py)

**Comment by kalvdans** [comment]
I'm a huge fan of disconnecting the select query creation from the session, I hope we could make this to work :)

File: app/backend/src/couchers/couchers_select.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/couchers_select.py]: Makes way more sense. Done.

File: app/backend/src/couchers/models.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/models.py]: LMK what you think of the new method

Comment (kalvdans) [reply app/backend/src/couchers/models.py]: hmm, was this property unused? Even easier :+1: 

I like it a lot that models.py does not deal with any session now.

Comment (ShmuelTreiger) [reply app/backend/src/couchers/models.py]: It was used, but only in conversations, which is where I put it.

Comment (ShmuelTreiger) [reply app/backend/src/couchers/models.py]: >I like it a lot that models.py does not deal with any session now.

Ya, that was a cool added benefit

File: app/backend/src/couchers/couchers_select.py
Comment (aapeliv) [inline app/backend/src/couchers/couchers_select.py]: Maybe we could call this file `couchers/sql.py`?

**Comment by ShmuelTreiger** [comment]
The only potential issue I know of regarding this update/migration is that it seems relationships still use Query objects, so if we ever want to do something like `cluster.members.where_users_visible()` we may run into some trouble. Other than that, this is ready to go.

**Comment by ShmuelTreiger** [comment]
@aapeliv @kalvdans any chance of getting this merged today?

File: app/backend/src/couchers/jobs/handlers.py
Comment (aapeliv) [inline app/backend/src/couchers/jobs/handlers.py]: Is this in `.sql`?

File: app/backend/src/couchers/jobs/handlers.py
Comment (aapeliv) [inline app/backend/src/couchers/jobs/handlers.py:16]: This seems to be imported everywhere like this. Should we instead just rename it as `select` in `couchers.sql`. I understand the thinking here, but I feel importing `select` directly would be better probably?

File: app/backend/src/couchers/jobs/handlers.py
Comment (aapeliv) [inline app/backend/src/couchers/jobs/handlers.py:234]: What does the `scalars()` do in these queries? I'd think it'd return either a single row, e.g. a single `User` (equivalent to `one()` before?) or then just one column (e.g. the primary key?)?

File: app/backend/src/couchers/jobs/worker.py
Comment (aapeliv) [inline app/backend/src/couchers/jobs/worker.py]: I think the old wording is better, but up to you

File: app/backend/src/couchers/servicers/api.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/api.py:769]: Not important, but did anything change here? Why did the formatter break this line?

File: app/backend/src/couchers/servicers/events.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/events.py]: Are we normally importing these from `.sql `too?

File: app/backend/src/couchers/servicers/groups.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/groups.py]: import from `.sql`?

File: app/backend/src/couchers/servicers/references.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/references.py]: from `.sql`?

File: app/backend/src/couchers/servicers/references.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/references.py]: Is `.c[0]` the first column? Are you able to do something like naming it (I think it's `label('whatever')`?), then `.c.whatever`?

File: app/backend/src/couchers/servicers/requests.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/requests.py:159]: You can refactor these to user `first`, but not critical

File: app/backend/src/couchers/servicers/requests.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/requests.py:441]: I'm not sure what's going on here. But I think this RPC is being cut anyway.

File: app/backend/src/couchers/servicers/threads.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/threads.py:41]: Is a `select_from` the recommended way to do a `count`? Is it just shorthand for a subquery? Just wondering.

File: app/backend/src/tests/test_auth.py
Comment (aapeliv) [inline app/backend/src/tests/test_auth.py]: sql?

**Review by aapeliv — COMMENTED** [review]
Yes, very close to merging. We can definitely push it out in the next couple of hours (though I'm in UTC+3 now so going to sleep in ~3 hours).

**Comment by aapeliv** [comment]
@ShmuelTreiger: there is an option to squash these commits into one so the sqlalchemy so the history is a bit cleaner with this big change. Not necessary to do it, but it might be a good idea if you'd like

File: app/backend/src/couchers/jobs/handlers.py
Comment (kalvdans) [reply app/backend/src/couchers/jobs/handlers.py:16]: I think being explicit using couchers_select everywhere makes it easier to read the  code without having to scroll up.

File: app/backend/src/couchers/jobs/handlers.py
Comment (kalvdans) [reply app/backend/src/couchers/jobs/handlers.py]: No, it is in sqlalchemy? I don't understand your comment.

Comment (ShmuelTreiger) [reply app/backend/src/couchers/jobs/handlers.py]: He wants me to import it from `sqlalchemy.sql`. No big deal, sure

File: app/backend/src/couchers/jobs/handlers.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/jobs/handlers.py:16]: I chose to name it `couchers_select` instead of `select` so that it's clear to a future developer we are using our own construct, and so that they don't accidentally import normal `select` and wonder why it's not working

I chose to rename it `select` so that our sql statements are as similar to normal SQLA syntax as possible, 1) for clarity and 2) for troubleshooting

Essentially I wanted it to look like normal SQLA, but I wanted to leave easy breadcrumbs so that some dev in 5 years can pick it up and figure it out

File: app/backend/src/couchers/jobs/handlers.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/jobs/handlers.py:234]: `.scalars()` returns an SQLA scalar object, which contains objects of the model type. So here, once you append `.all()`, it gives you a list of all the `User` objects.

https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=scalars#sqlalchemy.engine.Result.scalars

https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=scalars#sqlalchemy.engine.ScalarResult

File: app/backend/src/couchers/servicers/api.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/servicers/api.py:769]: Looks like nothing changed. I'm not sure why it did

File: app/backend/src/couchers/servicers/requests.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/servicers/requests.py:159]: `scalar_one` raises an exception if more than one is detected (or none), same as `one`. I didn't really do any reimagining of queries, just translation. Happy to change this and any others to `.first` if that's more fitting

File: app/backend/src/couchers/servicers/requests.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/servicers/requests.py:441]: I interpreted it as .one_or_none() would raise an exception if the wrong result was found, so I added an explicit abort. Couldn't figure out what else this was doing here.

File: app/backend/src/couchers/servicers/threads.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/servicers/threads.py:41]: This is their recommended format for count. Here's one place they mention it, though I definitely saw it somewhere else as well:

https://github.com/sqlalchemy/sqlalchemy/discussions/6334#:~:text=Correct%20-%20use%20select(func.count()).select_from(Pets)%20for%20that.

**Comment by ShmuelTreiger** [comment]
>@ShmuelTreiger: there is an option to squash these commits into one so the sqlalchemy so the history is a bit cleaner with this big change. Not necessary to do it, but it might be a good idea if you'd like

Either way is fine. I personally appreciate the history to see how things develop, but the squashing makes it cleaner for future users, so I see both ways.

File: app/backend/src/couchers/servicers/references.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/servicers/references.py]: Yes, it's the first column

`union_all` returns a tuple of values, no objects, hence the need for aliased. I can perhaps add a line to add a name? It doesn't add much, but it makes it a little clearer:

```
            union = union_all(q1, q2).order_by(HostRequest.end_time_to_write_reference.asc()).subquery()
            union = select(union.c[0].label("surfed"), aliased(HostRequest, union))
            host_request_references = session.execute(union).all()
```

File: app/backend/src/couchers/servicers/threads.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/servicers/threads.py:41]: I believe this is where I first got it:

https://stackoverflow.com/questions/65754023/what-is-the-equivalent-of-query-count-in-the-sqlalchemy-1-4-orm

Here it is in the docs:

https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#:~:text=Another%20example%20where%20we%20might%20want%20to%20use

File: app/backend/src/couchers/jobs/handlers.py
Comment (aapeliv) [reply app/backend/src/couchers/jobs/handlers.py:16]: OK, if you both think it's better, let's do this!

File: app/backend/src/couchers/jobs/handlers.py
Comment (aapeliv) [reply app/backend/src/couchers/jobs/handlers.py:234]: Thanks!

File: app/backend/src/couchers/servicers/api.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/api.py:769]: Fair enough. Was wondering if your formatter config was wrong (e.g. line length not 120). Seems it's an isolated thing though, so no problem

File: app/backend/src/couchers/servicers/references.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/references.py]: Yeah I think that's a bit clearer.

Also, what's the `aliased` doing here? I'm not sure I understand

File: app/backend/src/couchers/servicers/requests.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/requests.py:159]: That's fine

Comment (aapeliv) [reply app/backend/src/couchers/servicers/requests.py:159]: I didn't realise that difference!

File: app/backend/src/couchers/servicers/threads.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/threads.py:41]: Thanks!

**Comment by aapeliv** [comment]
I'm happy for this to be merged if you add that one `label` as in the remaining comment, and if you fix the tests!

> > @ShmuelTreiger: there is an option to squash these commits into one so the sqlalchemy so the history is a bit cleaner with this big change. Not necessary to do it, but it might be a good idea if you'd like
> 
> Either way is fine. I personally appreciate the history to see how things develop, but the squashing makes it cleaner for future users, so I see both ways.

Up to you!

File: app/backend/src/couchers/servicers/references.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/servicers/references.py]: It reconstitutes the object

https://github.com/sqlalchemy/sqlalchemy/discussions/6714

File: app/backend/src/couchers/jobs/handlers.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/jobs/handlers.py:16]: I think @kalvdans means to just import couchers_select and do session.execute(couchers_select(User))

So there's three options

File: app/backend/src/couchers/servicers/api.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/servicers/api.py:769]: image

🤷

File: app/backend/src/couchers/servicers/references.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/references.py]: Thanks!

**Review by aapeliv — APPROVED** [review]
LGTM if tests go green!

Once we merge, I'll push to next and then we'll have some people manually testing a bit just to make sure none of the more exotic parts are failing.

File: app/backend/src/couchers/jobs/handlers.py
Comment (ShmuelTreiger) [reply app/backend/src/couchers/jobs/handlers.py:16]: We never really resolved this. But we can discuss and if we want to rename, it's no big deal

Comment (aapeliv) [reply app/backend/src/couchers/jobs/handlers.py:16]: Could you remind us to discuss it in the next backend meeting?

**Pull request merged**
